### ISSUE TITLE
Chunking fails when chunked file in subdir

### ISSUE THREAD

**Issue by robander** [dita-ot/dita-ot#2706]
My map is: `sample/upandover.ditamap`

It references files outside of the `sample` directory, so there is extra "uplevels" processing. The map also has this reference for chunking:
`<topicref href="subdir/forchunks.dita" chunk="by-topic"/>`

The build crashes with the following error: 
```
chunk:
    [chunk] Processing file:/C:/DITA-OT/dita-ot-2.4.6/test/P025007/temp1/sample/upandover.ditamap
    [chunk] Processing file:/C:/DITA-OT/dita-ot-2.4.6/test/P025007/temp1/sample/subdir/forchunks.dita
    [chunk] C:\DITA-OT\dita-ot-2.4.6\test\P025007\temp1\sample\sample\subchunk.dita (The system cannot find the path specified.)
Error: java.util.NoSuchElementException
```

When I use a map that has the same references (but no uplevels), the build completes. However, the root chunk from `subdir/forchunks.dita` is missing; nothing is generated, and the `index.html` file has a link to `subdir/forchunks.dita.chunk` instead of to something in the map's directory as in reported in #2655 -- there are also a bunch of errors for that missing topic.

Attaching my sample files, the batch script (Windows) used to run, and the logs. I'm building `input.ditamap` (no uplevels), and `updandover.ditamap` with both `generate.copy.outer=1` and `generate.copy.outer=3`. The up and over test fails for each case. Test run with 2.4.6 but results match `develop` code for 2.5.
[chunkcrash.zip](https://github.com/dita-ot/dita-ot/files/1030514/chunkcrash.zip)

**Comment by robander** [comment]
Actually it looks like chunking `by-topic` is broken for any case in 2.4.6 + current `develop` code. With this very basic map:
```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE map PUBLIC "-//OASIS//DTD DITA Map//EN" "map.dtd">
<map xml:lang="en-us"><title>Something</title>
<topicref href="chunkme.dita" chunk="by-topic"/>
</map>
```

And this very simple topic:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE topic PUBLIC "-//OASIS//DTD DITA Topic//EN" "topic.dtd">
<topic id="r1">
<title>Root topic</title>
<body>
<p>Topic paragraph</p>
</body>
<topic id="c1">
<title>Child topic</title>
<body>
<p>hello there</p>
</body>
</topic>
</topic>
```

Building to HTML5, the root chunk is ignored. I only get the output file `c1.html` for the child. The root topic is referenced from the map as `chunkme.dita.chunk`, which doesn't exist, leading to errors like:
```
     [xslt] file:/C:/Users/IBM_ADMIN/github/robander/dita-ot/src/main/test/P0250
07/justchunk.ditamap:4:49: [DOTX009W][WARN]: Could not retrieve a title from 'ch
unkme.dita.chunk'. Using 'chunkme.dita.chunk' instead.
```

I'd really be quite happy if somebody could point to a simple user error on my part at this point...
[basic.zip](https://github.com/dita-ot/dita-ot/files/1030597/basic.zip)

**Comment by mironovalexey** [comment]
Cannot reproduce the latest case.
[justchunk.tar.gz](https://github.com/dita-ot/dita-ot/files/1030763/justchunk.tar.gz)

**Comment by robander** [comment]
Odd ... I'm still getting this on Windows, with latest code from `release/2.5` branch. Here's a zip of the source + log + output: 
[morechunk.zip](https://github.com/dita-ot/dita-ot/files/1039444/morechunk.zip)

I wonder if this could be an operating system thing - I've long gotten failures from `./gradlew test` on Windows, all from chunk modules, which do not cause problems on Linux. I haven't managed to diagnose them. I don't remember when they started appearing. I thought it had to do with test cases and did not reflect a real problem. I just went back to version 2.3.1 with the same test files, and got the same bad results.

**Comment by robander** [comment]
I can't imagine anybody wants to diagnose these right now, but FWIW, here is what I get from running `gradlew test` on Windows - have gotten the same 14 "failures" for quite a while now.
```
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:compileTestJava UP-TO-DATE
:processTestResources UP-TO-DATE
:testClasses UP-TO-DATE
:test

org.dita.dost.module.ChunkMapReaderTest > testconflict_by_topic FAILED
    java.lang.AssertionError at ChunkMapReaderTest.java:1126

org.dita.dost.module.ChunkMapReaderTest > testAttribute_map10 FAILED
    java.lang.AssertionError at ChunkMapReaderTest.java:1126

org.dita.dost.module.ChunkMapReaderTest > testAttribute_map11 FAILED
    java.lang.AssertionError at ChunkMapReaderTest.java:1126

org.dita.dost.module.ChunkMapReaderTest > testByTopic_map2 FAILED
    java.lang.AssertionError at ChunkMapReaderTest.java:1126

org.dita.dost.module.ChunkMapReaderTest > testByTopic_map7 FAILED
    java.lang.AssertionError at ChunkMapReaderTest.java:1126

org.dita.dost.module.ChunkMapReaderTest > testcopy_to1 FAILED
    java.lang.RuntimeException at ChunkMapReaderTest.java:1129
        Caused by: java.lang.NullPointerException at ChunkMapReaderTest.java:1124

org.dita.dost.module.ChunkMapReaderTest > testcopy_to2 FAILED
    java.lang.RuntimeException at ChunkMapReaderTest.java:1129
        Caused by: java.lang.NullPointerException at ChunkMapReaderTest.java:1124

org.dita.dost.module.ChunkModuleTest > testByTopicMap2 FAILED
    junit.framework.AssertionFailedError at ChunkModuleTest.java:177

org.dita.dost.module.ChunkModuleTest > testByTopicMap7 FAILED
    junit.framework.AssertionFailedError at ChunkModuleTest.java:202

org.dita.dost.module.ChunkModuleTest > testConflictByTopic FAILED
    junit.framework.AssertionFailedError at ChunkModuleTest.java:70

org.dita.dost.module.ChunkModuleTest > testCopyTo1 FAILED
    java.lang.RuntimeException at ChunkModuleTest.java:75
        Caused by: java.lang.NullPointerException at ChunkModuleTest.java:75

org.dita.dost.module.ChunkModuleTest > testCopyTo2 FAILED
    java.lang.RuntimeException at ChunkModuleTest.java:80
        Caused by: java.lang.NullPointerException at ChunkModuleTest.java:80

org.dita.dost.module.ChunkModuleTest > testAttributeMap10 FAILED
    java.lang.RuntimeException at ChunkModuleTest.java:167
        Caused by: java.lang.RuntimeException at ChunkModuleTest.java:167
            Caused by: java.io.FileNotFoundException at ChunkModuleTest.java:167

org.dita.dost.module.ChunkModuleTest > testAttributeMap11 FAILED
    junit.framework.AssertionFailedError at ChunkModuleTest.java:172
432 tests completed, 14 failed, 8 skipped
:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///C:/Users/IBM_ADMIN/github
/dita-ot/build/reports/tests/test/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug
option to get more log output.

BUILD FAILED

Total time: 2 mins 25.531 secs
```

**Comment by robander** [comment]
Update: I downloaded DITA-OT 2.5 on a Windows 10 machine that isn't used for development, and reproduced the issue there, with same errors / results. So it fails on my dev system (Windows 7 + IBM Java) and on another test system (Windows 10 + Oracle Java). So, current guess is that this is somehow a Windows issue, which is ... oddd.

**Comment by robander** [comment]
OK, tracked down the issue, and found 2 potential fixes. Not sure which we should use (maybe both).

The issue is [here](https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/writer/SeparateChunkTopicParser.java#L113):
`firstTopicID = getFirstTopicId(currentFile.resolve(parseFilePath).getPath());`

On Windows, my test file is sending this in as the resolved path:
`/C:/Users/IBM_ADMIN/github/robander/dita-ot/src/main/test/chunk/temp/chunkme.dita`

The `getFirstTopicID` method [begins with](https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/writer/AbstractChunkTopicParser.java#L419):
```
    String getFirstTopicId(final String absolutePathToFile) {
        assert new File(absolutePathToFile).isAbsolute();
        if (!isAbsolutePath(absolutePathToFile)) {
            return null;
        }
```

The `new File(absolutePathToFile).isAbsolute()` test returns true - Java thinks this is an absolute path.

The `isAbsolutePath(absolutePathToFile)` returns false - our own utility function thinks this is not an absolute path.

I've tried two different fixes that each work (and bonus, they fix the 14 chunk tests that have been failing on my system for a year or more):
- Swap out our `isAbsolutePath` test and test `if (new File(absolutePathToFile).isAbsolute())`
- Update our `isAbsolutePath` [utility function](https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/util/FileUtils.java#L428) to allow for `/C:/` (any drive letter) at the start of a path:
`return (path.matches("([a-zA-Z]:|\\\\)\\\\.*") || path.matches("/[a-zA-Z]:/.*"));`

A better fix, as Jarno just suggested elsewhere, is to have the `getFirstTopicID` function take in a File to begin with, rather than converting a URI to a string and passing the string. Working on that now.

**Comment by robander** [comment]
Resolved with #2728.
